### PR TITLE
feat(math): implement packed field extension

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This adds Extension Field of the Packed Field.

### CODE REVIEW

**Review by ashjeong — COMMENTED** [review]
https://github.com/kroma-network/tachyon/pull/478/commits/6e337aed6a2e590a2ff97b6efedd3cc9d9e0d53b
Please change the commit title to be: "change constexpr variables to static member variables"
Fix the typo of “refactoring” in the commit message

https://github.com/kroma-network/tachyon/pull/478/commits/5b5fcbcfb62ae018c1affc6ef227b11baf33c704
Fix the typo in the commit title : "initialization"

**Comment by TomTaehoonKim** [comment]
@chokobole Is this typo or did you intend? 0453008, rerepresent -> represent.

**Review by TomTaehoonKim — APPROVED** [review]
LGTM. Can you add reference of `non_residue`s in the commit message body?

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by ashjeong — APPROVED** [review]
LGTM

File: tachyon/math/finite_fields/koala_bear/internal/BUILD.bazel [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    class_name = "KoalaBear4",
+    is_packed = False,
+    namespace = "tachyon::math",
+    non_residue = ["3"],
+    deps = ["//tachyon/math/finite_fields/koala_bear"],
+)
+
+generate_fp4s(
+    name = "packed_koala_bear4",
+    base_field = "PackeKoalaBear",
```
Comment (batzor) [inline tachyon/math/finite_fields/koala_bear/internal/BUILD.bazel:20]:
```suggestion
    base_field = "PackedKoalaBear",
```

File: tachyon/math/finite_fields/koala_bear/internal/BUILD.bazel [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    base_field_degree = 1,
+    base_field_hdr = "tachyon/math/finite_fields/koala_bear/packed_koala_bear.h",
+    class_name = "PackeKoalaBear4",
```
Comment (batzor) [inline tachyon/math/finite_fields/koala_bear/internal/BUILD.bazel:23]:
```suggestion
    class_name = "PackedKoalaBear4",
```

File: tachyon/math/finite_fields/koala_bear/BUILD.bazel [hunk 3]
```diff
@@ -34,6 +34,18 @@ generate_fp4s(
     deps = [":koala_bear"],
 )
 
+generate_fp4s(
+    name = "packed_koala_bear4",
+    base_field = "PackeKoalaBear",
```
Comment (batzor) [inline tachyon/math/finite_fields/koala_bear/BUILD.bazel]:
```suggestion
    base_field = "PackedKoalaBear",
```

File: tachyon/math/finite_fields/koala_bear/BUILD.bazel [hunk 4]
```diff
@@ -34,6 +34,18 @@ generate_fp4s(
...
+    base_field_degree = 1,
+    base_field_hdr = "tachyon/math/finite_fields/koala_bear/packed_koala_bear.h",
+    class_name = "PackeKoalaBear4",
```
Comment (batzor) [inline tachyon/math/finite_fields/koala_bear/BUILD.bazel]:
```suggestion
    class_name = "PackedKoalaBear4",
```

**Review by GideokKim — APPROVED** [review]
LGTM

**Comment by ashjeong** [comment]
Should the `FromBasePrimeFields()` functions in quadratic, cubic, and quartic extension fields also be changed to `FromBaseFields()`?

**Comment by chokobole** [comment]
> Should the `FromBasePrimeFields()` functions in quadratic, cubic, and quartic extension fields also be changed to `FromBaseFields()`?

Exactly! `GenerateInitExtField()` should also be changed, but since they are not used, so I'll just leave a comment!

File: tachyon/math/finite_fields/fp3.h [hunk 5]
```diff
@@ -41,13 +43,17 @@ class Fp3 final : public CubicExtensionField<Fp3<Config>> {
     //    = α₀ + α₁(x³)^((P - 1) / 3) * x + α₂(x³)^(2 * (P - 1) / 3) * x²
     //    = α₀ + α₁ωx + α₂ω²x², where ω is a cubic root of unity.
 
-    constexpr size_t N = BasePrimeField::kLimbNums;
+    using UnpackedBasePrimeField = std::conditional_t<
+        FiniteFieldTraits<BasePrimeField>::kIsPackedPrimeField,
+        typename FiniteFieldTraits<BasePrimeField>::PrimeField, BasePrimeField>;
```
Comment (dongchangYoo) [inline tachyon/math/finite_fields/fp3.h:48]: How about implementing this code in FiniteFieldTraits? It appears repeatedly.

**Review by dongchangYoo — CHANGES_REQUESTED** [review]

**Review by batzor — APPROVED** [review]
LGTM

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Pull request merged**